signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt

```diff
@@ -0,0 +1,214 @@
+         */
+        fun BigInteger.getUnsignedBytes(): ByteArray {
+            val bytes = this.toByteArray()
+            if (bytes[0] == 0x0.toByte()) {
+                return bytes.drop(1).toByteArray()
+            }
+            return bytes
+        }
+
+        // Canonicalize - In order to remove malleability,
+        // we set s = curve_order - s, if s is greater than curve.Order() / 2.
+        var sigS = s
+        if (sigS > halfCurveOrder) {
+            sigS = curve.n.subtract(sigS)
+        }
+
+        val sBytes = sigS.getUnsignedBytes()
+        val rBytes = r.getUnsignedBytes()
+
+        require(rBytes.size <= 32) { "cannot encode r into BTC Format, size overflow (${rBytes.size} > 32)" }
+        require(sBytes.size <= 32) { "cannot encode s into BTC Format, size overflow (${sBytes.size} > 32)" }
+
+        val signature = ByteArray(64)
+        // 0 pad the byte arrays from the left if they aren't big enough.
+        System.arraycopy(rBytes, 0, signature, 32 - rBytes.size, rBytes.size)
+        System.arraycopy(sBytes, 0, signature, 64 - sBytes.size, sBytes.size)
+
+        return BTCSignature.fromByteArray(signature)
+    }
+}
+
+/**
+ * ASN.1-specific signature format operations.
+ */
+object ASN1 {
+    /**
+     * Encodes the `(r,s)` pair as an ASN.1 DER byte array, suitable for use with the Java [JavaSignature] API
+     * for signature verification.
+     *
+     * As an ASN.1 structure, the ECDSA signature is defined as
+     *
+     * ```
+     * ECDSASignature ::= SEQUENCE {
+     *     r   INTEGER,
+     *     s   INTEGER
+     * }
+     * ```
+     *
+     * A usage example:
+     *
+     * ```
+     * val privateKey: PrivateKey = ...
+     * val payload: ByteArray = ... // raw array of bytes
+     * val payloadHash: ByteArray = payloadHash.sha256()
+     * val signature: EDCSASignature = BCECSigner().sign(privateKey, payloadHash)
+     * val ok: Boolean = Signature.getInstance("SHA256withECDSA").apply {
+     *     initVerify(publicKey)
+     *     update(payload)
+     *     s.verify(signature.encodeAsASN1DER())
+     * }
+     * assert(ok)
+     * ```
+     *
+     * @param r The `r` [BigInteger] to encode.
+     * @param s The `s` [BigInteger] to encode.
+     * @return `(r,s)` pair encoded as a ASN.1 DER array of bytes.
+     */
+    fun encode(r: BigInteger, s: BigInteger): ASN1Signature =
+        ByteArrayOutputStream().use {
+            DERSequenceGenerator(it).apply {
+                addObject(ASN1Integer(r.toByteArray()))
+                addObject(ASN1Integer(s.toByteArray()))
+                close()
+            }
+            ASN1Signature.fromByteArray(it.toByteArray())
+        }
+
+    /**
+     * Decode an ASN.1 signature into an `(r,s)` integer pair.
+     *
+     * @param bytes The signature bytes to decode.
+     * @return The decoded `(r,s)` integer pair as a [BigIntegerPair].
+     */
+    fun decode(bytes: ByteArray): BigIntegerPair =
```

Comment:
future note: If BigIntegerPair is converted to a data class, callers can use decomposition to pull the r and s values directly as `val (r, s) = ASN1.decode(bytes)`

Reply:
That's a great point